### DISCUSSION TITLE
/collection/hashmap.html

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#38] [Announcements]
# /collection/hashmap.html

一份通俗易懂、风趣幽默的Java学习指南，内容涵盖Java基础、Java并发编程、Java虚拟机、Java企业级开发、Java面试等核心知识点。学Java，就认准Java程序员进阶之路

https://tobebetterjavaer.com/collection/hashmap.html

**Comment by xiaoacaia** [comment]
$$
a ~ \% ~ b = a ~ \& ~ (b-1) \\
a ~ \% ~ 2^n = a ~ \& ~ (2^n-1) \\
比如：14 ~ \% ~ 8 = 14 ~ \& ~ 7 \\
\because 14 = 1110, ~ 7 = 0111 \\
\therefore 14 ~ \& ~ 7 = 0110 = 6
$$

**Comment by xiaoacaia** [comment]
$$
a ~ \% ~ b = a ~ \& ~ (b-1) \\
a ~ \% ~ 2^n = a ~ \& ~ (2^n-1) \\
比如：14 ~ \% ~ 8 = 14 ~ \& ~ 7 \\
\because 14 = 1110, ~ 7 = 0111 \\
\therefore 14 ~ \& ~ 7 = 0110 = 6
$$

**Reply by xiaoacaia** [reply]
评论不能删除吗。。 那个 & 、% 前需要加 \ 转义

**Reply by xiaoacaia** [reply]
(n - 1) & hash  =>  (h = key.hashCode()) ^ (h >>> 16) （保证均匀性）  &  n = tab.length（高位舍去操作）

**Reply by xiaoacaia** [reply]
(n - 1) & hash  =>  (h = key.hashCode()) ^ (h >>> 16) （混合高位信息，增大随机性）  &  n = tab.length（高位舍去操作）
